### DISCUSSION TITLE
Upload PDF on Product DataObject edition page

### DISCUSSION THREAD

**Discussion by julescournut** [pimcore/pimcore#11019] [Q&A / Support] [answered]
Hello everyone !

I'm currently working on a Pimcore 10.2.5 and wish to have a PDF upload field in my Product DataObject. 
I'm using a link for the moment with my pdf in the Assets treeview but to add a new pdf to a product, it's quite complicated because you have to upload a file in assets then find it and link it to your product. 
I rather have an upload field directly on my product edition page, same as the one for images (Image DataType) but that would work with pdf. 
Is it possible and not too complicated to create such a field ? 

Alternatively how can I order my assets in treeview by desc date instead of alphabetical order ? So that if you import a new pdf it comes out in first position.
![Capture d’écran 2021-12-14 à 14 23 30](https://user-images.githubusercontent.com/26310511/146012329-badd32f4-2539-4c5a-bb6e-5554670616c9.png)

Thanks for your answers

**Comment by NiklasBr** [comment, marked as answer]
In your class definition you can declare a target path where assets are uploaded when uploaded directly on the dataobject:

![image](https://user-images.githubusercontent.com/279826/146018746-329b5287-a304-4e00-86b7-c43c88bc0dda.png)

And on the rightmost part of the field you should have an upload icon (it does not have a tooltip/title) to explain what it does, but it works by helping you upload files directly:

![image](https://user-images.githubusercontent.com/279826/146019149-6990ee71-c2a9-43e9-a88f-2850d2392cf4.png)

**Reply by julescournut** [reply]
Thank you for your answer,
Which Data Component are you using to have this field ? I couldn't find this "Assets Restrictions" configuration page anywhere

**Reply by NiklasBr** [reply]
Relation types (Many-to-many, adv. many-to-many, many-to-one, ...)

**Reply by julescournut** [reply]
Thanks a lot that's exactly what I wanted !

**Comment by IvanDotsenko** [comment]
We at IQMosaic have developed a Document module where you can upload any type of document to Data Objects as easily as images.
Available for PIMCore 11.

https://github.com/iqmosaic/pimcore-admin-ui-bundle


https://github.com/pimcore/pimcore/assets/18754070/1e31a986-c881-4ef0-8e9a-80f94bfd6f02

**Comment by idna001** [comment]
@IvanDotsenko Nice, is there a change to change the MIME Type, would like to upload a vcf file directly in a Data Object. 
I tried your plugin, but I'm getting this error, as it is no PDF: 
<img width="705" alt="grafik" src="https://github.com/pimcore/pimcore/assets/46223359/d218b4fd-79a5-4e57-aa4d-1118d22c75f8">

I also tried removing and change your if check about it, but it ran in some other exception. 
Thanks for a tip

**Reply by IvanDotsenko** [reply]
@idna001 
This error is related to the default Pimcore document type setting. You can easily fix it by editing your project's config file (`config/config.yaml`) and adding the following additional configuration:

```
pimcore:
    assets:
        type_definitions:
            map:
                document:
                    class: \Pimcore\Model\Asset\Document
                    matching: ['/msword/', '/pdf/', '/powerpoint/', '/office/', '/excel/', '/opendocument/', '/vcf/']
```

**Reply by idna001** [reply]
Wow, perfect, thanks for your fast help. 
This works. 👍🏼

**Reply by IvanDotsenko** [reply]
Also if you want to have a document gallery/list you can use `Field Collections `

<img width="496" alt="image" src="https://github.com/pimcore/pimcore/assets/18754070/733a6a93-4988-46a2-b099-5e776e829916">

**Reply by idna001** [reply]
Hi again, 
I'm still using this. But now I want to add a file via PHP API. Can you explain what I missed here in the following?
```					$vcfDocument = new \Pimcore\Model\Asset\Document();
                    $vcfDocument->setParentId(11);
                    $vcfDocument->setFilename(date('Y-m-d_H-i-s'). '-'. $vcf->getClientOriginalName());
                    $vcfDocument->setData(file_get_contents($vcfPath));
                    $vcfDocument->save();
                    $member->setUploadVcf($vcfDocument);
```

Usually I would create just an Asset, not Asset\Document(); 
But the method requires such. 
Also, there is no error message, the asset is uploading correctly. But the link to the data object is not set. 

Thanks for a hint.
Andreas

**Reply by IvanDotsenko** [reply]
@idna001 I didn't test it with the PHP API. I need to check it.

**Reply by NiklasBr** [reply]
@idna001 did you save `$member` with `->save()` to persist the relationship?

**Reply by idna001** [reply]
I tested this again and found my error in the meantime. My testfile had an unsupported file type.. 
I will add an exception to create an error message for my case.

Thanks,  @IvanDotsenko and @NiklasBr for your fast response 😄
